lib/semantics/resolve-names.cc

```diff
@@ -5915,6 +5917,318 @@ class DeferredCheckVisitor {
+  switch (beginDir.v) {
+  case parser::OmpLoopDirective::Directive::Distribute: {
+    PushContext(beginDir.source, OmpDirective::DISTRIBUTE);
+  } break;
+  case parser::OmpLoopDirective::Directive::Do: {
+    PushContext(beginDir.source, OmpDirective::DO);
+  } break;
+  case parser::OmpLoopDirective::Directive::DoSimd: {
+    PushContext(beginDir.source, OmpDirective::DO_SIMD);
+  } break;
+  case parser::OmpLoopDirective::Directive::ParallelDo: {
+    PushContext(beginDir.source, OmpDirective::PARALLEL_DO);
+  } break;
+  case parser::OmpLoopDirective::Directive::ParallelDoSimd: {
+    PushContext(beginDir.source, OmpDirective::PARALLEL_DO_SIMD);
+  } break;
+  case parser::OmpLoopDirective::Directive::Simd: {
+    PushContext(beginDir.source, OmpDirective::SIMD);
+  } break;
+  case parser::OmpLoopDirective::Directive::Taskloop: {
+    PushContext(beginDir.source, OmpDirective::TASKLOOP);
+  } break;
+  case parser::OmpLoopDirective::Directive::TaskloopSimd: {
+    PushContext(beginDir.source, OmpDirective::TASKLOOP_SIMD);
+  } break;
+  default:
+    // TODO others
+    break;
+  }
+  ClearDataSharingAttributeObjects();
+  return true;
+}
+
+bool OmpAttributeVisitor::Pre(const parser::OpenMPSectionsConstruct &x) {
+  const auto &beginSectionsDir{
+      std::get<parser::OmpBeginSectionsDirective>(x.t)};
+  const auto &beginDir{
+      std::get<parser::OmpSectionsDirective>(beginSectionsDir.t)};
+  switch (beginDir.v) {
+  case parser::OmpSectionsDirective::Directive::ParallelSections: {
+    PushContext(beginDir.source, OmpDirective::PARALLEL_SECTIONS);
+  } break;
+  case parser::OmpSectionsDirective::Directive::Sections: {
+    PushContext(beginDir.source, OmpDirective::SECTIONS);
+  } break;
+  default: break;
+  }
+  ClearDataSharingAttributeObjects();
+  return true;
+}
+
+bool OmpAttributeVisitor::Pre(const parser::OpenMPThreadprivate &x) {
+  PushContext(x.source, OmpDirective::THREADPRIVATE);
+  const auto &list{std::get<parser::OmpObjectList>(x.t)};
+  ResolveOmpObjectList(list, Symbol::Flag::OmpThreadprivate);
+  return false;
+}
+
+void OmpAttributeVisitor::Post(const parser::OmpDefaultClause &x) {
+  if (!ompContext_.empty()) {
+    switch (x.v) {
+    case parser::OmpDefaultClause::Type::Private:
+      SetContextDefaultDSA(Symbol::Flag::OmpPrivate);
+      break;
+    case parser::OmpDefaultClause::Type::Firstprivate:
+      SetContextDefaultDSA(Symbol::Flag::OmpFirstPrivate);
+      break;
+    case parser::OmpDefaultClause::Type::Shared:
+      SetContextDefaultDSA(Symbol::Flag::OmpShared);
+      break;
+    case parser::OmpDefaultClause::Type::None:
+      SetContextDefaultDSA(Symbol::Flag::OmpNone);
+      break;
+    }
+  }
+}
+
+// For OpenMP constructs, check all the data-refs within the constructs
+// and adjust the symbol for each Name if necessary
+void OmpAttributeVisitor::Post(const parser::Name &name) {
+  if (!ompContext_.empty() && GetContext().withinConstruct) {
```

Comment:
It is intended for now because it falls into the TODO category for the `implicitly determined` data-sharing attributes, which is listed in the `OpenMP-semantics.md` doc. In this case when processing `sections`, we need to locate the upper `parallel` context and see what the `default` DSA is. I am currently working on the nested constructs example.